server/src/main/java/org/opensearch/indices/IndicesRequestCache.java

```diff
@@ -228,19 +240,31 @@ BytesReference getOrCompute(
         final Key key = new Key(((IndexShard) cacheEntity.getCacheIdentity()).shardId(), cacheKey, readerCacheKeyId);
         Loader cacheLoader = new Loader(cacheEntity, loader);
         BytesReference value = cache.computeIfAbsent(key, cacheLoader);
-        if (cacheLoader.isLoaded()) {
-            cacheEntity.onMiss();
-            // see if it's the first time we see this reader, and make sure to register a cleanup key
-            CleanupKey cleanupKey = new CleanupKey(cacheEntity, readerCacheKeyId);
-            if (!registeredClosedListeners.containsKey(cleanupKey)) {
-                Boolean previous = registeredClosedListeners.putIfAbsent(cleanupKey, Boolean.TRUE);
-                if (previous == null) {
-                    OpenSearchDirectoryReader.addReaderCloseListener(reader, cleanupKey);
+        CleanupKey cleanupKey = null;
+        try {
+            if (cacheLoader.isLoaded()) {
+                cacheEntity.onMiss();
+                // see if it's the first time we see this reader, and make sure to register a cleanup key
+                cleanupKey = new CleanupKey(cacheEntity, readerCacheKeyId);
+                if (!registeredClosedListeners.containsKey(cleanupKey)) {
+                    Boolean previous = registeredClosedListeners.putIfAbsent(cleanupKey, Boolean.TRUE);
+                    if (previous == null) {
+                        OpenSearchDirectoryReader.addReaderCloseListener(reader, cleanupKey);
+                    }
                 }
+                cacheCleanupManager.updateCleanupKeyToCountMapOnCacheInsertion(cleanupKey);
+            } else {
+                cacheEntity.onHit();
+            }
+        } catch (Exception e) {
```

Comment:
Instead of introducing try-catch block at this level, should we just wrap
```OpenSearchDirectoryReader.addReaderCloseListener(reader, cleanupKey);``` with try catch block?

Also should we be catching specific exception in this case? AutoCloseException or whatever that is?
As in this case if cache threw exception for other reasons(like internal etc), key may not be cached necessarily but we still end up enqueuing it for clean up.

Reply:
Wrt to your comments on 
>Instead of introducing try-catch block at this level, should we just wrap
OpenSearchDirectoryReader.addReaderCloseListener(reader, cleanupKey); with try catch block?

and 

>As in this case if cache threw exception for other reasons(like internal etc), key may not be cached necessarily but we still end up enqueuing it for clean up.

Any failures, after caching the entry would result in the entry being stale. 
That's why the try block encovers all the operations after the cache entry. That way any failures regardless of cache or not would generate an exception and hence cleaning the stale entry. 

This is also the reason for catching a general Exception over AlreadyClosedException.

Reply:
But cache can also throw exception before caching an entry? How would you know that?
One scenario: if the entry is not present in cache, its value is loaded by running the actual query which can itself result in exception and in this case entry was not cached.

Reply:
if the cache throws an exception while caching an entry, it raises an exception and we throw the execption back to the caller, following the original behavior. 
What i have done here is to catch any exceptions after caching only for addressing stale key and cleanups and throw it back.

Reply:
Due to this https://github.com/opensearch-project/OpenSearch/pull/13070#discussion_r1561757421
This entire thread is stale 😄